ocaml/testsuite/tests/typing-layouts/allow-illegal-crossing.ml

```diff
@@ -0,0 +1,654 @@
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type t is value, because
+         it's a boxed record type.
+       But the layout of type t must be a sublayout of value, because
+         of the annotation on the declaration of the type t.
+|}]
+
+type ('a, 'b) t : value mod many = { a : 'a; b : 'b }
+[%%expect {|
+Line 1, characters 0-53:
+1 | type ('a, 'b) t : value mod many = { a : 'a; b : 'b }
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type t is value, because
+         it's a boxed record type.
+       But the layout of type t must be a sublayout of value, because
+         of the annotation on the declaration of the type t.
+|}]
+
+type a : value mod unique = private b
+and b : value
+[%%expect {|
+Line 1, characters 0-37:
+1 | type a : value mod unique = private b
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type b is value, because
+         of the annotation on the declaration of the type b.
+       But the layout of type b must be a sublayout of value, because
+         of the definition of a at line 1, characters 0-37.
+|}]
+
+type a : value
+and b : value mod global = private a
+[%%expect {|
+Line 2, characters 0-36:
+2 | and b : value mod global = private a
+    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type a/3 is value, because
+         of the annotation on the declaration of the type a.
+       But the layout of type a/3 must be a sublayout of value, because
+         of the definition of b at line 2, characters 0-36.
+|}]
+
+module rec A : sig
+  type t : value mod external_
+end = struct
+  type t : value mod external_ = private B.t
+end
+and B : sig
+  type t : value
+end = struct
+  type t : value
+end
+[%%expect {|
+Line 4, characters 2-44:
+4 |   type t : value mod external_ = private B.t
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type B.t is value, because
+         of the definition of t at line 7, characters 2-16.
+       But the layout of type B.t must be a sublayout of immediate, because
+         of the definition of t at line 4, characters 2-44.
+|}]
+
+module rec A : sig
+  type t : value
+end = struct
+  type t : value
+end
+and B : sig
+  type t : value mod many
+end = struct
+  type t : value mod many = private A.t
+end
+[%%expect {|
+Line 9, characters 2-39:
+9 |   type t : value mod many = private A.t
+      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+Error: The layout of type A.t is value, because
+         of the definition of t at line 2, characters 2-16.
+       But the layout of type A.t must be a sublayout of value, because
+         of the definition of t at line 9, characters 2-39.
+|}]
+
+(*********************************************************************************)
+(* Test 4: types cannot cross portability and contention axes when not annotated *)
+
+(* CR: should this be allowed? I think no, the user should have to specify at the
```

Comment:
Do whatever is easiest. I agree that rejecting is probably better. But accepting this is very much in the spirit of this patch, and it may be hard to reject.